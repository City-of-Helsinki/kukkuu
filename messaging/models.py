from random import random

from django.db import models
from django.utils.timezone import now
from django.utils.translation import gettext_lazy as _
from parler.models import TranslatedFields
from projects.models import Project

from common.models import TimestampedModel, TranslatableModel, TranslatableQuerySet
from events.models import Event, Occurrence
from users.models import Guardian


class MessageQuerySet(TranslatableQuerySet):
    def user_can_view(self, user):
        return self.filter(project__users=user)


class Message(TimestampedModel, TranslatableModel):
    ALL = "all"
    INVITED = "invited"
    ENROLLED = "enrolled"
    ATTENDED = "attended"
    SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION = "subscribed_to_free_spot_notification"
    RECIPIENT_SELECTION_CHOICES = (
        (ALL, _("All")),
        (INVITED, _("Invited")),
        (ENROLLED, _("Enrolled")),
        (ATTENDED, _("Attended")),
        (
            SUBSCRIBED_TO_FREE_SPOT_NOTIFICATION,
            _("Subscribed to free spot notification"),
        ),
    )

    project = models.ForeignKey(
        Project,
        verbose_name=_("project"),
        related_name="messages",
        on_delete=models.CASCADE,
    )
    translations = TranslatedFields(
        subject=models.CharField(verbose_name=_("subject"), max_length=255),
        body_text=models.TextField(verbose_name=_("body plain text")),
    )
    sent_at = models.DateTimeField(verbose_name=_("sent at"), blank=True, null=True)
    recipient_count = models.PositiveIntegerField(
        verbose_name=_("recipient count"), blank=True, null=True
    )
    recipient_selection = models.CharField(
        max_length=64,
        verbose_name=_("recipient selection"),
        choices=RECIPIENT_SELECTION_CHOICES,
    )
    event = models.ForeignKey(
        Event,
        verbose_name=_("event"),
        related_name="messages",
        blank=True,
        null=True,
        on_delete=models.CASCADE,
    )
    occurrences = models.ManyToManyField(
        Occurrence, verbose_name=_("occurrences"), related_name="messages", blank=True,
    )

    objects = MessageQuerySet.as_manager()

    class Meta:
        ordering = ("id",)
        verbose_name = _("message")
        verbose_name_plural = _("messages")

    def __str__(self):
        return f"({self.pk}) {self.subject} ({self.sent_at or 'not sent'})"

    def send(self):
        self.recipient_count = (
            Guardian.objects.filter(children__project=self.project).count()
            if self.recipient_selection == Message.ALL
            else random.randint(
                0, 50
            )  # TODO temporary thingy before actual send implementation
        )
        self.sent_at = now()
        self.save(update_fields=("recipient_count", "sent_at"))

    def can_user_administer(self, user):
        return user.projects.filter(pk=self.project.pk).exists()
