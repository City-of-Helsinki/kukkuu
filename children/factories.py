import factory

from children.models import Child, Relationship
from projects.models import Project
from users.factories import GuardianFactory


class ChildFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    birthyear = factory.Faker(
        "random_int", min=2018, max=2023
    )  # NOTE: Many of the tests are freezed to 2020
    postal_code = factory.Faker("postcode", locale="fi_FI")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))
    notes = factory.Faker(
        "random_element",
        elements=[
            "",
            "Test notes",
            "Alternative test notes",
            "Longer test notes\nwith multiple lines.",
        ],
    )

    class Meta:
        model = Child


class RelationshipFactory(factory.django.DjangoModelFactory):
    child = factory.SubFactory(ChildFactory)
    guardian = factory.SubFactory(GuardianFactory)
    type = factory.Faker(
        "random_element", elements=[t[0] for t in Relationship.TYPE_CHOICES]
    )

    class Meta:
        model = Relationship


class ChildWithGuardianFactory(ChildFactory):
    relationship = factory.RelatedFactory(RelationshipFactory, "child")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))


class ChildWithTwoGuardiansFactory(ChildWithGuardianFactory):
    relationship2 = factory.RelatedFactory(RelationshipFactory, "child")
