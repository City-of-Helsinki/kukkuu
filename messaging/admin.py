from django.contrib import admin
from django.utils.translation import gettext_lazy as _
from parler.admin import TranslatableAdmin

from .models import AlreadySentError, Message


@admin.register(Message)
class MessageAdmin(TranslatableAdmin):
    list_display = (
        "subject",
        "project",
        "sent_at",
        "get_recipient_count",
        "created_at",
        "updated_at",
    )
    fields = (
        "project",
        "subject",
        "body_text",
        "recipient_selection",
        "event",
        "occurrences",
        "created_at",
        "updated_at",
        "get_recipient_count",
    )
    readonly_fields = ("get_recipient_count", "created_at", "updated_at")
    actions = ("send",)

    def get_recipient_count(self, obj):
        return str(obj.get_recipient_count()) + (
            str(_(" (if sent now)")) if not obj.sent_at else ""
        )

    get_recipient_count.short_description = _("recipient count")

    def send(self, request, queryset):
        sent_count = 0
        for obj in queryset:
            try:
                obj.send()
                sent_count += 1
            except AlreadySentError:
                pass
        self.message_user(request, _("%s successfully sent.") % sent_count)
