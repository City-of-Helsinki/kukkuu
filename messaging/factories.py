import factory
from projects.models import Project

from .models import Message


class MessageFactory(factory.django.DjangoModelFactory):
    subject = factory.Faker("text", max_nb_chars=20)
    body_text = factory.Faker("paragraph", nb_sentences=5)
    project = factory.LazyFunction(lambda: Project.objects.get_or_create(year=2020)[0])
    recipient_selection = Message.ALL

    class Meta:
        model = Message
