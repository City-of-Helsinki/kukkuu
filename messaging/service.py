from typing import TYPE_CHECKING

from django.conf import settings
from django.utils import timezone
from parler.utils.context import switch_language

from django_ilmoitin.utils import Message as MailerMessage
from django_ilmoitin.utils import send_all, send_mail
from kukkuu.notification_service import send_sms_notification
from messaging.exceptions import AlreadySentError
from users.models import Guardian

if TYPE_CHECKING:
    from messaging.models import Message


def send_message(message: "Message", *, force=False):
    from messaging.models import Message

    if message.sent_at and not force:
        raise AlreadySentError()

    guardians = message.get_recipient_guardians()

    message.sent_at = timezone.now()
    message.recipient_count = len(guardians)

    message.save(update_fields=("sent_at", "recipient_count"))

    for guardian in guardians:
        with switch_language(message, guardian.language):
            if message.protocol == Message.EMAIL:
                send_email_notification(guardian, message)
            elif message.protocol == Message.SMS:
                send_sms_notification([guardian.phone_number], message.body_text)

    if not getattr(settings, "ILMOITIN_QUEUE_NOTIFICATIONS", False):
        MailerMessage.objects.retry_deferred()
        send_all()


def send_email_notification(guardian: Guardian, message: "Message"):
    # hopefully this functionality that uses django-ilmoitin's internals
    # is only temporarily here, and will be removed when either
    # 1) we need to use a notification template and thus will use
    #    ilmoitin's regular send_notification(), or
    # 2) support for sending a mail without a notification template will
    #    be added to ilmoitin and we can use that
    if guardian.language in getattr(settings, "ILMOITIN_TRANSLATED_FROM_EMAIL", {}):
        from_email = settings.ILMOITIN_TRANSLATED_FROM_EMAIL[guardian.language]
    else:
        from_email = settings.DEFAULT_FROM_EMAIL

    send_mail(
        message.subject,
        message.body_text,
        guardian.email,
        from_email=from_email,
    )
