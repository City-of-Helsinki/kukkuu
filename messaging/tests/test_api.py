from datetime import datetime

import pytest
from django.core import mail
from django.utils.timezone import now
from guardian.shortcuts import assign_perm
from messaging.factories import MessageFactory
from messaging.models import Message

from children.factories import ChildWithGuardianFactory
from common.tests.utils import assert_match_error_code, assert_permission_denied
from common.utils import get_global_id
from events.factories import EventFactory, OccurrenceFactory
from kukkuu.consts import MESSAGE_ALREADY_SENT_ERROR, OBJECT_DOES_NOT_EXIST_ERROR

MESSAGES_QUERY = """
query Messages {
  messages {
    edges {
      node {
        project {
          year
        }
        subject
        bodyText
        recipientSelection
        sentAt
        recipientCount
        event {
          name
        }
        occurrences {
          edges {
            node {
              time
            }
          }
        }
      }
    }
  }
}
"""


def test_messages_query(snapshot, project_user_api_client, message, another_project):
    MessageFactory(project=another_project)

    executed = project_user_api_client.execute(MESSAGES_QUERY)

    snapshot.assert_match(executed)


def test_cannot_do_messages_query_unauthorized(user_api_client, message):
    executed = user_api_client.execute(MESSAGES_QUERY)

    assert_permission_denied(executed)


def test_messages_query_project_filter(
    snapshot, project_user_api_client, message, project, another_project
):
    assign_perm("admin", project_user_api_client.user, another_project)

    executed = project_user_api_client.execute(
        MESSAGES_QUERY, variables={"project_id": get_global_id(project)}
    )

    snapshot.assert_match(executed)


MESSAGE_QUERY = """
query Message($id: ID!) {
  message(id: $id){
    project {
      year
    }
    subject
    bodyText
    recipientSelection
    sentAt
    recipientCount
    event {
      name
    }
    occurrences {
      edges {
        node {
          time
        }
      }
    }
  }
}
"""


def test_message_query(snapshot, project_user_api_client, message):
    executed = project_user_api_client.execute(
        MESSAGE_QUERY, variables={"id": get_global_id(message)}
    )

    snapshot.assert_match(executed)


def test_cannot_do_message_query_unauthorized(user_api_client, message):
    executed = user_api_client.execute(
        MESSAGE_QUERY, variables={"id": get_global_id(message)}
    )

    assert_permission_denied(executed)


def test_cannot_do_message_query_unauthorized_wrong_project(
    snapshot, wrong_project_api_client, message
):
    executed = wrong_project_api_client.execute(
        MESSAGE_QUERY, variables={"id": get_global_id(message)}
    )

    snapshot.assert_match(executed)


ADD_MESSAGE_MUTATION = """
mutation AddMessage($input: AddMessageMutationInput!) {
  addMessage(input: $input) {
    message {
      translations {
        languageCode
        subject
        bodyText
      }
      sentAt
      recipientSelection
      recipientCount
      project {
        year
      }
      event {
        name
      }
      occurrences {
        edges {
          node {
            time
          }
        }
      }
    }
  }
}
"""


def get_add_message_variables(project, event=None, occurrences=None):
    variables = {
        "input": {
            "translations": [
                {
                    "subject": "Testiotsikko",
                    "bodyText": "Testiteksti",
                    "languageCode": "FI",
                }
            ],
            "recipientSelection": "ALL",
            "projectId": get_global_id(project),
        }
    }
    if event:
        variables["input"]["eventId"] = get_global_id(event)
    if occurrences is not None:
        variables["input"]["occurrenceIds"] = [get_global_id(o) for o in occurrences]
    return variables


@pytest.mark.parametrize("event_selection", (None, "event", "occurrences"))
@pytest.mark.django_db
def test_add_message(snapshot, project_user_api_client, project, event_selection):
    if event_selection == "event":
        variables = get_add_message_variables(
            project, event=EventFactory(published_at=now())
        )
    elif event_selection == "occurrences":
        event = EventFactory(published_at=now())
        occurrences = OccurrenceFactory.create_batch(3, event=event)
        variables = get_add_message_variables(
            project, event=event, occurrences=occurrences
        )
    else:
        variables = get_add_message_variables(project)

    executed = project_user_api_client.execute(
        ADD_MESSAGE_MUTATION, variables=variables
    )

    snapshot.assert_match(executed)


@pytest.mark.django_db
def test_cannot_add_message_unauthorized(project, wrong_project_api_client):
    executed = wrong_project_api_client.execute(
        ADD_MESSAGE_MUTATION, variables=get_add_message_variables(project)
    )

    assert_permission_denied(executed)


UPDATE_MESSAGE_MUTATION = """
mutation UpdateMessage($input: UpdateMessageMutationInput!) {
  updateMessage(input: $input) {
    message {
      translations {
        languageCode
        subject
        bodyText
      }
      sentAt
      recipientSelection
      recipientCount
      project {
        year
      }
      event {
        name
      }
      occurrences {
        edges {
          node {
            time
          }
        }
      }
    }
  }
}
"""


def get_update_message_variables(message, event=None, occurrences=None):
    variables = {
        "input": {
            "translations": [
                {
                    "subject": "P채ivitetty testiotsikko",
                    "bodyText": "P채ivitetty testiteksti.",
                    "languageCode": "FI",
                }
            ],
            "recipientSelection": "ATTENDED",
            "id": get_global_id(message),
        }
    }
    if event:
        variables["input"]["eventId"] = get_global_id(event)
    if occurrences is not None:
        variables["input"]["occurrenceIds"] = [get_global_id(o) for o in occurrences]
    return variables


@pytest.mark.parametrize("event_selection", (None, "event", "event_and_occurrences"))
@pytest.mark.django_db
def test_update_message(snapshot, project_user_api_client, event_selection):
    old_event = EventFactory(published_at=now())
    message = MessageFactory(event=old_event)
    old_occurrences = OccurrenceFactory.create_batch(2, event=message.event)
    message.occurrences.set(old_occurrences)

    if event_selection == "event":
        new_event = EventFactory(published_at=now())
        new_occurrences = []
    elif event_selection == "event_and_occurrences":
        new_event = EventFactory(published_at=now())
        new_occurrences = [
            OccurrenceFactory(
                event=new_event,
                time=datetime(2016, 8, 16, 7, 10, 0, tzinfo=now().tzinfo),
            )
        ]
    else:
        new_event = None
        new_occurrences = []

    variables = get_update_message_variables(
        message, event=new_event, occurrences=new_occurrences
    )

    executed = project_user_api_client.execute(
        UPDATE_MESSAGE_MUTATION, variables=variables
    )

    snapshot.assert_match(executed)
    message.refresh_from_db()
    assert message.subject == "P채ivitetty testiotsikko"
    assert message.body_text == "P채ivitetty testiteksti."
    assert message.event == (
        new_event
        if event_selection in ("event", "event_and_occurrences")
        else old_event
    )
    assert [o.pk for o in message.occurrences.all()] == [o.pk for o in new_occurrences]


@pytest.mark.django_db
def test_cannot_update_message_unauthorized(message, wrong_project_api_client):
    executed = wrong_project_api_client.execute(
        UPDATE_MESSAGE_MUTATION, variables=get_update_message_variables(message),
    )

    assert_match_error_code(executed, OBJECT_DOES_NOT_EXIST_ERROR)


@pytest.mark.django_db
def test_cannot_update_sent_message(project_user_api_client, sent_message):
    executed = project_user_api_client.execute(
        UPDATE_MESSAGE_MUTATION,
        variables={"input": {"id": get_global_id(sent_message)}},
    )
    assert_match_error_code(executed, MESSAGE_ALREADY_SENT_ERROR)


SEND_MESSAGE_MUTATION = """
mutation SendMessage($input: SendMessageMutationInput!) {
  sendMessage(input: $input) {
    message {
      subject
      sentAt
      recipientCount
    }
  }
}
"""


@pytest.mark.django_db
def test_send_message(snapshot, project_user_api_client, message):
    ChildWithGuardianFactory()

    executed = project_user_api_client.execute(
        SEND_MESSAGE_MUTATION, variables={"input": {"id": get_global_id(message)}}
    )

    snapshot.assert_match(executed)
    assert len(mail.outbox) == 1


@pytest.mark.django_db
def test_cannot_send_message_unauthorized(wrong_project_api_client, message):
    executed = wrong_project_api_client.execute(
        SEND_MESSAGE_MUTATION, variables={"input": {"id": get_global_id(message)}}
    )

    assert_match_error_code(executed, OBJECT_DOES_NOT_EXIST_ERROR)


@pytest.mark.django_db
def test_cannot_send_message_more_than_once(project_user_api_client, sent_message):
    ChildWithGuardianFactory()

    executed = project_user_api_client.execute(
        SEND_MESSAGE_MUTATION, variables={"input": {"id": get_global_id(sent_message)}}
    )

    assert_match_error_code(executed, MESSAGE_ALREADY_SENT_ERROR)
    assert len(mail.outbox) == 0


DELETE_MESSAGE_MUTATION = """
mutation DeleteMessage($input: DeleteMessageMutationInput!) {
   deleteMessage(input: $input) {
    clientMutationId
  }
}
"""


@pytest.mark.django_db
def test_delete_message(snapshot, project_user_api_client, message):
    executed = project_user_api_client.execute(
        DELETE_MESSAGE_MUTATION, variables={"input": {"id": get_global_id(message)}}
    )

    snapshot.assert_match(executed)
    assert Message.objects.count() == 0


@pytest.mark.django_db
def test_cannot_delete_message_unauthorized(wrong_project_api_client, message):
    executed = wrong_project_api_client.execute(
        DELETE_MESSAGE_MUTATION, variables={"input": {"id": get_global_id(message)}}
    )

    assert_match_error_code(executed, OBJECT_DOES_NOT_EXIST_ERROR)


@pytest.mark.django_db
def test_cannot_delete_sent_message(project_user_api_client, sent_message):
    executed = project_user_api_client.execute(
        DELETE_MESSAGE_MUTATION,
        variables={"input": {"id": get_global_id(sent_message)}},
    )

    assert_match_error_code(executed, MESSAGE_ALREADY_SENT_ERROR)
    assert Message.objects.count() == 1
