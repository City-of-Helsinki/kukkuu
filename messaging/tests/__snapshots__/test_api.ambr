# serializer version: 1
# name: test_add_message[None]
  dict({
    'data': dict({
      'addMessage': dict({
        'message': dict({
          'event': None,
          'occurrences': dict({
            'edges': list([
            ]),
          }),
          'project': dict({
            'year': 2020,
          }),
          'protocol': 'EMAIL',
          'recipientCount': 0,
          'recipientSelection': 'ALL',
          'sentAt': None,
          'translations': list([
            dict({
              'bodyText': 'Testiteksti',
              'languageCode': 'FI',
              'subject': 'Testiotsikko',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_add_message[event]
  dict({
    'data': dict({
      'addMessage': dict({
        'message': dict({
          'event': dict({
            'name': 'Poor lawyer treat free heart significant.',
          }),
          'occurrences': dict({
            'edges': list([
            ]),
          }),
          'project': dict({
            'year': 2020,
          }),
          'protocol': 'EMAIL',
          'recipientCount': 0,
          'recipientSelection': 'ALL',
          'sentAt': None,
          'translations': list([
            dict({
              'bodyText': 'Testiteksti',
              'languageCode': 'FI',
              'subject': 'Testiotsikko',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_add_message[occurrences]
  dict({
    'data': dict({
      'addMessage': dict({
        'message': dict({
          'event': dict({
            'name': 'Poor lawyer treat free heart significant.',
          }),
          'occurrences': dict({
            'edges': list([
              dict({
                'node': dict({
                  'time': '1970-12-29T14:07:50.629900+00:00',
                }),
              }),
              dict({
                'node': dict({
                  'time': '1977-02-25T22:54:59.889967+00:00',
                }),
              }),
              dict({
                'node': dict({
                  'time': '1997-09-11T00:12:17.610651+00:00',
                }),
              }),
            ]),
          }),
          'project': dict({
            'year': 2020,
          }),
          'protocol': 'EMAIL',
          'recipientCount': 0,
          'recipientSelection': 'ALL',
          'sentAt': None,
          'translations': list([
            dict({
              'bodyText': 'Testiteksti',
              'languageCode': 'FI',
              'subject': 'Testiotsikko',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_cannot_do_message_query_unauthorized_wrong_project
  dict({
    'data': dict({
      'message': None,
    }),
  })
# ---
# name: test_delete_message
  dict({
    'data': dict({
      'deleteMessage': dict({
        'clientMutationId': None,
      }),
    }),
  })
# ---
# name: test_message_query
  dict({
    'data': dict({
      'message': dict({
        'bodyText': 'Ruumisteksti.',
        'event': None,
        'occurrences': dict({
          'edges': list([
          ]),
        }),
        'project': dict({
          'year': 2020,
        }),
        'recipientCount': 0,
        'recipientSelection': 'ALL',
        'sentAt': None,
        'subject': 'Otsikko',
      }),
    }),
  })
# ---
# name: test_messages_query
  dict({
    'data': dict({
      'messages': dict({
        'count': 1,
        'edges': list([
          dict({
            'node': dict({
              'bodyText': 'Ruumisteksti.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Otsikko',
            }),
          }),
        ]),
        'pageInfo': dict({
          'endCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
          'hasNextPage': False,
          'hasPreviousPage': False,
          'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
        }),
      }),
    }),
  })
# ---
# name: test_messages_query_occurrences_filter
  dict({
    'data': dict({
      'messages': dict({
        'count': 4,
        'edges': list([
          dict({
            'node': dict({
              'bodyText': 'Free heart significant machine try. President compare room hotel town south among. Fall long respond draw military dog. Increase thank certainly again thought summer. Beyond than trial western.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                  dict({
                    'node': dict({
                      'time': '1980-08-23T04:26:09.363315+00:00',
                    }),
                  }),
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Him question stay.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Child care any. Minute defense level church. Alone our very television beat at success.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                  dict({
                    'node': dict({
                      'time': '1980-08-23T04:26:09.363315+00:00',
                    }),
                  }),
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Business hot PM.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Focus significant kind. Laugh smile behavior whom gas. Significant minute rest. Special far magazine.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                  dict({
                    'node': dict({
                      'time': '2003-01-02T17:50:46.571751+00:00',
                    }),
                  }),
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Attention practice.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Conference carry factor front Mr amount conference thing. Positive cold start rest tonight including believe. Respond range bit college question. Stop treatment suggest. Sometimes growth check court.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                  dict({
                    'node': dict({
                      'time': '2003-01-02T17:50:46.571751+00:00',
                    }),
                  }),
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Past life thus.',
            }),
          }),
        ]),
        'pageInfo': dict({
          'endCursor': 'YXJyYXljb25uZWN0aW9uOjM=',
          'hasNextPage': False,
          'hasPreviousPage': False,
          'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
        }),
      }),
    }),
  })
# ---
# name: test_messages_query_project_filter
  dict({
    'data': dict({
      'messages': dict({
        'count': 1,
        'edges': list([
          dict({
            'node': dict({
              'bodyText': 'Ruumisteksti.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Otsikko',
            }),
          }),
        ]),
        'pageInfo': dict({
          'endCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
          'hasNextPage': False,
          'hasPreviousPage': False,
          'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
        }),
      }),
    }),
  })
# ---
# name: test_messages_query_protocol_filter[email]
  dict({
    'data': dict({
      'messages': dict({
        'count': 5,
        'edges': list([
          dict({
            'node': dict({
              'bodyText': 'Which president smile staff country actually generation. Age member whatever open effort clear. Difficult look can. Care figure mention wrong when lead involve. Event lay yes policy data control as receive.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Partner exist true.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Girl middle same space speak. Person the probably deep center develop character situation. Score think turn argue present.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Decade every town.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Voice radio happen color scene. Create state rock only. Several behavior media career decide season mission TV. Work head table city central deep response. Through resource professional debate produce college able.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Walk fish teach.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Sea something western research. Candidate century network bar hear quite wonder. Up always sport return. Light a point charge stand store. Generation able take food share.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Product issue along.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Site chance of performance. Hand cause receive kitchen middle. Step able last in able local garden. Nearly gun two born land military first.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'EMAIL',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Discussion remain.',
            }),
          }),
        ]),
        'pageInfo': dict({
          'endCursor': 'YXJyYXljb25uZWN0aW9uOjQ=',
          'hasNextPage': False,
          'hasPreviousPage': False,
          'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
        }),
      }),
    }),
  })
# ---
# name: test_messages_query_protocol_filter[sms]
  dict({
    'data': dict({
      'messages': dict({
        'count': 5,
        'edges': list([
          dict({
            'node': dict({
              'bodyText': 'Free heart significant machine try. President compare room hotel town south among. Fall long respond draw military dog. Increase thank certainly again thought summer. Beyond than trial western.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'SMS',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Him question stay.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Child care any. Minute defense level church. Alone our very television beat at success.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'SMS',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Business hot PM.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Enter everything history remember stay public high. Exist shoulder write century. Never skill down subject town. According hard enough watch condition like lay.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'SMS',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Detail.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'From daughter order stay sign discover eight. Toward scientist service wonder everything. Middle moment strong hand push book and interesting. Near increase process truth list pressure. Capital city sing himself yard stuff.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'SMS',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Their tell.',
            }),
          }),
          dict({
            'node': dict({
              'bodyText': 'Base may middle good father boy economy. Fly discussion huge get this success. Science sort already name. Senior number scene today friend maintain marriage.',
              'event': None,
              'occurrences': dict({
                'edges': list([
                ]),
              }),
              'project': dict({
                'year': 2020,
              }),
              'protocol': 'SMS',
              'recipientCount': 0,
              'recipientSelection': 'ALL',
              'sentAt': None,
              'subject': 'Even perhaps that.',
            }),
          }),
        ]),
        'pageInfo': dict({
          'endCursor': 'YXJyYXljb25uZWN0aW9uOjQ=',
          'hasNextPage': False,
          'hasPreviousPage': False,
          'startCursor': 'YXJyYXljb25uZWN0aW9uOjA=',
        }),
      }),
    }),
  })
# ---
# name: test_send_message
  dict({
    'data': dict({
      'sendMessage': dict({
        'message': dict({
          'protocol': 'EMAIL',
          'recipientCount': 1,
          'sentAt': '2020-12-12T00:00:00+00:00',
          'subject': 'Otsikko',
        }),
      }),
    }),
  })
# ---
# name: test_send_sms_message_sent_with_default_language
  dict({
    'data': dict({
      'sendMessage': dict({
        'message': dict({
          'protocol': 'SMS',
          'recipientCount': 1,
          'sentAt': '2020-12-12T00:00:00+00:00',
          'subject': 'Otsikko',
        }),
      }),
    }),
  })
# ---
# name: test_update_message[None]
  dict({
    'data': dict({
      'updateMessage': dict({
        'message': dict({
          'event': dict({
            'name': 'Poor lawyer treat free heart significant.',
          }),
          'occurrences': dict({
            'edges': list([
            ]),
          }),
          'project': dict({
            'year': 2020,
          }),
          'protocol': 'SMS',
          'recipientCount': 0,
          'recipientSelection': 'ATTENDED',
          'sentAt': None,
          'translations': list([
            dict({
              'bodyText': 'Päivitetty testiteksti.',
              'languageCode': 'FI',
              'subject': 'Päivitetty testiotsikko',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_update_message[event]
  dict({
    'data': dict({
      'updateMessage': dict({
        'message': dict({
          'event': dict({
            'name': 'Up property work indeed take. Popular care receive camera.',
          }),
          'occurrences': dict({
            'edges': list([
            ]),
          }),
          'project': dict({
            'year': 2020,
          }),
          'protocol': 'SMS',
          'recipientCount': 0,
          'recipientSelection': 'ATTENDED',
          'sentAt': None,
          'translations': list([
            dict({
              'bodyText': 'Päivitetty testiteksti.',
              'languageCode': 'FI',
              'subject': 'Päivitetty testiotsikko',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_update_message[event_and_occurrences]
  dict({
    'data': dict({
      'updateMessage': dict({
        'message': dict({
          'event': dict({
            'name': 'Up property work indeed take. Popular care receive camera.',
          }),
          'occurrences': dict({
            'edges': list([
              dict({
                'node': dict({
                  'time': '2016-08-16T07:10:00+00:00',
                }),
              }),
            ]),
          }),
          'project': dict({
            'year': 2020,
          }),
          'protocol': 'SMS',
          'recipientCount': 0,
          'recipientSelection': 'ATTENDED',
          'sentAt': None,
          'translations': list([
            dict({
              'bodyText': 'Päivitetty testiteksti.',
              'languageCode': 'FI',
              'subject': 'Päivitetty testiotsikko',
            }),
          ]),
        }),
      }),
    }),
  })
# ---
